### ISSUE TITLE
[Itweaker] How could i do my ASM Work before loading Mixins?

### ISSUE THREAD

**Issue by kingdevnl** [SpongePowered/Mixin#208]
Im currenyly having an issue that Mixin overwrites my changes that i made using ASM

This its my injectIntoClassLoader method inside the Iweaker i have
```java
    @Override
    public void injectIntoClassLoader(LaunchClassLoader classLoader) {

        classLoader.addTransformerExclusion("com.enderloader.EnderLoader");
        classLoader.addTransformerExclusion("com.enderloader.asm.transformers.AccesModder");

        classLoader.registerTransformer("com.enderloader.asm.transformers.AccesModder");

        MixinBootstrap.init();

        Mixins.addConfiguration("mixins.client.json");

        MixinEnvironment.getDefaultEnvironment().setSide(MixinEnvironment.Side.CLIENT);

    }
```

**Comment by kingdevnl** [comment]
PS: I have to do the Changes with ASM before Mixins 
Like (Acces Transforming of fields etc)

**Comment by Mumfrey** [comment]
You can make any changes you like with transformers, however mixin will nearly always apply its changes last in order to be working with the bytecode in its final form. Mixin will not overwrite access changes to fields and it will only overwrite methods when they are specifically decorated with `@Overwrite` annotations.

When an overwrite occurs, the original method is gone and there is no way to transform it. These overwrites are performed last because the replaced method may change significantly from the original and so should not be transformed as if they are the original method.

This is all by design, but there are defined ways to work reliably with mixin code which are provided to you. The options are:

* **Use mixin yourself to _interact_ with the _other mixins_**. All mixins registered in the environment are marshalled according to their priority, this means you can usurp another mixin in a controlled way by supplying *your own* mixin and then using your own `@Overwrite` to overwrite the changes made by the first mixin.
  
  Alternatively you can supply an *empty* mixin for the class in question, but provide a *companion plugin* for your mixin config which will then receive [postApply](https://github.com/SpongePowered/Mixin/blob/master/src/main/java/org/spongepowered/asm/mixin/extensibility/IMixinConfigPlugin.java#L123) callbacks which allow you to apply transforms *after* the mixins are applied.

* **Have your transformer _transform the mixins_**. All mixins which are loaded are actually transformed so your transformer can actually alter them if it wants to. This is a better way to handle transformations in `@Overwrite` methods because you will know whether you are transforming the original method in the target class (eg. `WorldServer`) or an overwriting version of the method (which may be different) in the mixin (eg. `MixinWorldServer`).
  
  This approach is very powerful because transformers run on the mixin code *before* mixin does and processing on it, so it is possible to run any transforms you like and mixin will validate your changes and will fail with a useful error if something goes wrong.

So basically, your transformer can do anything it likes to the target class, and Mixin won't care, in fact it will automatically adapt to the changes. The only changes mixin will "undo" are when a mixin overwrites a specific method, but there are mechanisms already provided which allow you to interact with the custom code if you want to.

**Comment by Mumfrey** [comment]
Oh, I also realised from re-reading your issue that it might be your own mixins which are overwriting your own changes. In which case the same things still apply, they just describe ways to interact with your mixins which are possible using a companion plugin or in your existing transformer.

**Comment by kingdevnl** [comment]
Getting a Bit too advanced for me to get Haha, Thanks for the Fast reply aswell!

**Comment by kingdevnl** [comment]
What its the Best way to do it for a Newcomer to Mixin?

**Comment by Mumfrey** [comment]
Again it depends what you want to do. If you want to transform the mixin code, you can do that because the mixins pass through your transformer, otherwise you can use a config companion plugin and just use the `postApply` callback which receives the ASM tree node for the target class after all the mixins have been applied. It's really up to you.

Like I said, mixin won't ever "undo" your changes though, with the exception of `@Overwrite`, so those are the only cases you need to actually deal with.

**Comment by kingdevnl** [comment]
Im not using Overwrite anywhere only using the Inject system to Inject in the method init()V

**Comment by Mumfrey** [comment]
In which case mixin will not overwrite any of the changes made by your transformer.

**Comment by kingdevnl** [comment]
mmm, Hers a Link to my code https://github.com/EnderLoader/EnderLoader

**Comment by Mumfrey** [comment]
Your transformer changes the class bytes but never actually returns them: [AccesModder.java line 28](https://github.com/EnderLoader/EnderLoader/blob/1.12-DEV/src/main/java/com/enderloader/asm/transformers/AccesModder.java#L28).

Next time you should verify that your transformer works *without* Mixin before wondering if Mixin is the thing breaking it 😀
